src/compaction_filter.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include "db_impl.h"
+#include "rocksdb/compaction_filter.h"
+
+#include <utility>
+
+namespace rocksdb {
+namespace titandb {
+
+class TitanCompactionFilter final : public CompactionFilter {
+public:
+  explicit TitanCompactionFilter(const CompactionFilter *original,
+                                 std::weak_ptr<BlobStorage> blob_storage)
+      : blob_storage_(std::move(blob_storage)), original_filter_(original) {}
+
+  const char *Name() const override {
+    return std::string("TitanCompactionFilter.")
+        .append(original_filter_->Name())
+        .c_str();
+  }
+
+  Decision FilterV2(int level, const Slice &key, ValueType value_type,
+                    const Slice &value, std::string *new_value,
+                    std::string *skip_until) const override {
+    if (value_type != kBlobIndex) {
+      return original_filter_->FilterV2(level, key, value_type, value,
+                                        new_value, skip_until);
+    }
+
+    BlobIndex blob_index;
+    Slice original_value(value.data());
+    Status s = blob_index.DecodeFrom(&original_value);
+    if (!s.ok()) {
+      // Unable to decode blob index. Keeping the value.
+      return Decision::kKeep;
+    }
+    if (BlobIndex::IsDeletionMarker(blob_index)) {
+      // TODO(yiwu): handle deletion marker at bottom level.
+      return Decision::kKeep;
+    }
+
+    BlobRecord record;
+    PinnableSlice buffer;
+
+    auto storage = blob_storage_.lock();
+    if (storage) {
+      ReadOptions read_options;
+      s = storage->Get(read_options, blob_index, &record, &buffer);
+    } else {
+      // Column family not found, remove the value.
+      return Decision::kRemove;
+    }
+
+    if (s.ok()) {
+      auto decision = original_filter_->FilterV2(
+          level, key, ValueType::kValue, record.value, new_value, skip_until);
+      // It would be a problem if you change the value whereas the value_type is
+      // still kBlobIndex. For now, we can just assert value_change == false and
+      // abort.
+      // TODO: we should make rocksdb Filter API support changing value_type
+      assert(decision != CompactionFilter::Decision::kChangeValue);
```

Comment:
hmmm, we should at least handle this one. Let's return `kKeep` if user compaction filter request to change the value.

Reply:
but it maybe makes customized filter logic wrong and breaks data correctness.

Reply:
Agree. Let's set background error in this case then. Also set background error for blob index decode error above.

Reply:
```cpp
      if (decision == Decision::kChangeValue) {
        decision = Decision::kKeep;
      }
      return decision;
```